### PR TITLE
Upgrade to PHPStan 2

### PR DESCRIPTION
**Pull request by leofeyer**

### CODE REVIEW

File: core-bundle/tests/EventListener/DbafsMetadataSubscriberTest.php [hunk 1]
```diff
@@ -60,7 +60,7 @@ public function testEnhancesMetadata(): void
         $localizedMetadata = $extraMetadata->getLocalized();
 
         $this->assertInstanceOf(MetadataBag::class, $localizedMetadata);
-        $this->assertInstanceOf(Metadata::class, $localizedMetadata['de']);
+        $this->assertArrayHasKey('de', $localizedMetadata);
```
Comment (Toflar) [inline core-bundle/tests/EventListener/DbafsMetadataSubscriberTest.php:63]: This is not the same check.

File: core-bundle/tests/Translation/DataCollectorTranslatorTest.php [hunk 2]
```diff
@@ -98,7 +97,6 @@ public function testServiceIsResetable(): void
 
         $translator = new DataCollectorTranslator($originalTranslator);
 
-        $this->assertInstanceOf(ResetInterface::class, $translator);
```
Comment (Toflar) [inline core-bundle/tests/Translation/DataCollectorTranslatorTest.php:101]: Why this change?

File: core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php [hunk 3]
```diff
@@ -47,7 +47,6 @@ public function testCanReadTheOffset(): void
 
         $this->assertSame('bar', $bag['foo']);
         $this->assertSame(['baz' => 'foo'], $bag['bar']);
-        $this->assertSame('foo', $bag['bar']['baz']);
```
Comment (Toflar) [inline core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php:50]: Why?

File: core-bundle/tests/Routing/ResponseContext/CoreResponseContextFactoryTest.php [hunk 4]
```diff
@@ -65,9 +64,7 @@ public function testResponseContext(): void
             $this->createMock(UrlGeneratorInterface::class),
         );
 
-        $responseContext = $factory->createResponseContext();
-
-        $this->assertInstanceOf(ResponseHeaderBag::class, $responseContext->getHeaderBag());
```
Comment (Toflar) [inline core-bundle/tests/Routing/ResponseContext/CoreResponseContextFactoryTest.php:70]: This does not check anything anymore?

File: core-bundle/tests/Image/Studio/FigureBuilderTest.php [hunk 5]
```diff
@@ -679,9 +679,6 @@ public function testLastExceptionIsResetWhenCallingFrom(): void
 
         foreach ($setValidResourceOperations as [$method, $argument]) {
             $figureBuilder->from($invalidModel);
-
-            $this->assertInstanceOf(InvalidResourceException::class, $exception);
```
Comment (Toflar) [inline core-bundle/tests/Image/Studio/FigureBuilderTest.php:683]: Why?

File: core-bundle/tests/Image/ImageSizesTest.php [hunk 6]
```diff
@@ -128,8 +127,6 @@ public function testReturnsTheRegularUserOptions(): void
 
     public function testServiceIsResetable(): void
     {
-        $this->assertInstanceOf(ResetInterface::class, $this->imageSizes);
```
Comment (Toflar) [inline core-bundle/tests/Image/ImageSizesTest.php:131]: Why?

File: core-bundle/tests/Framework/ContaoFrameworkTest.php [hunk 7]
```diff
@@ -336,8 +335,6 @@ public function testRegistersTheHookServices(): void
 
     public function testServiceIsResetable(): void
     {
-        $this->assertInstanceOf(ResetInterface::class, $this->getFramework());
```
Comment (Toflar) [inline core-bundle/tests/Framework/ContaoFrameworkTest.php:339]: Why?

**Review by Toflar — CHANGES_REQUESTED** [review]
Most of it looks good but some of the changes in the tests look like the tests actually got worse?

File: core-bundle/tests/Framework/ContaoFrameworkTest.php (hunk 7 above)
Comment (leofeyer) [reply core-bundle/tests/Framework/ContaoFrameworkTest.php:339]:
```
Call to method PHPUnit\Framework\Assert::assertInstanceOf() with 'Symfony\\Contracts\\Service\\ResetInterface' and
Contao\CoreBundle\Framework\ContaoFramework will always evaluate to true.
🪪  method.alreadyNarrowedType
```

Because `ContaoFramework` implements `ResetInterface`.

File: core-bundle/tests/EventListener/DbafsMetadataSubscriberTest.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/tests/EventListener/DbafsMetadataSubscriberTest.php:63]: Correct. If `$localizedMetadata['de']` is set, it can never not be a `Metadata` class.

```
Call to method PHPUnit\Framework\Assert::assertInstanceOf() with 'Contao\\CoreBundle\\File\\Metadata' and Contao\CoreBundle\File\Metadata will always
evaluate to true.
🪪  method.alreadyNarrowedType
```

File: core-bundle/tests/Translation/DataCollectorTranslatorTest.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/tests/Translation/DataCollectorTranslatorTest.php:101]: See https://github.com/contao/contao/pull/7711#discussion_r1837898458

File: core-bundle/tests/Routing/ResponseContext/CoreResponseContextFactoryTest.php (hunk 4 above)
Comment (leofeyer) [reply core-bundle/tests/Routing/ResponseContext/CoreResponseContextFactoryTest.php:70]: See https://github.com/contao/contao/pull/7711#discussion_r1837898458

There is another expectation in line 52, so the check still checks something.

File: core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php (hunk 3 above)
Comment (leofeyer) [reply core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php:50]: Because of the assertion above. If `$bag['bar']` is `['baz' => 'foo']`, `$bag['bar']['baz']` can never be anything else but `'foo'`.

File: core-bundle/tests/Image/Studio/FigureBuilderTest.php (hunk 5 above)
Comment (leofeyer) [reply core-bundle/tests/Image/Studio/FigureBuilderTest.php:683]: See https://github.com/contao/contao/pull/7711#discussion_r1837898458

File: core-bundle/tests/Image/ImageSizesTest.php (hunk 6 above)
Comment (leofeyer) [reply core-bundle/tests/Image/ImageSizesTest.php:131]: See https://github.com/contao/contao/pull/7711#discussion_r1837898458

File: core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php (hunk 3 above)
Comment (m-vo) [reply core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php:50]: In general that's only true for arrays, though, not for classes implementing `ArrayAccess`.

**Review by Toflar — DISMISSED** [review]
I think this looks good, just wanted to double check those removed lines in the tests are correct :)

File: core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php (hunk 3 above)
Comment (leofeyer) [reply core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php:50]: Is it true in this specific case? Because if not, I can add the line again with a `@phpstan-ignore` comment.

Comment (m-vo) [reply core-bundle/tests/Session/Attribute/ArrayAttributeBagTest.php:50]: No, here it's fine to drop the line. :+1:
